client.go

```diff
@@ -25,44 +25,52 @@ var clientServerConnection net.Conn
 var clientMasterConnection rpc.Server
 //Need a map from clientId->serverConnPort
 
+// client LogicalClock
+var clientLogicalClock = 0
 
 type ClientMaster int
 
 type PutArgs struct {
 	key, value string
-	clientId, serverId int
+	clientId, serverId, clientLogicalClock int
 }
 
 type GetArgs struct {
 	key string
 }
 
+type history struct {
+	action string
+	key, value string
+	clientId, serverId int
+	timeStamp map[int]int
+}
+
+var clientHistory []history
+
 var clientWaitGroup sync.WaitGroup
 
 func (t *ClientMaster) clientPut(clientId int, key string, value string) error {
+	// Increment clients logical clock on receiving a put request from the master
+	clientLogicalClock += 1
 
-	/*
-	get serverId from clientId from map in master
-	get serverBasePort from serverId from map in master
-	*/
-
+	// Make a put request to the connected server
 	if serverConn == 0 {
 		return errors.New("connection does not exist")
 	}
-
- 	args := &PutArgs{key, value, clientId, serverConn}
- 	var reply int
-
+ 	args := &PutArgs{key, value, clientId, serverConn, clientLogicalClock}
+ 	// reply contains vectorTimeStamp corresponding to this transaction
+ 	var reply shared.Clock
 	client := rpc.NewClient(clientServerConnection)
-	err := client.Call("ClientServer.ServerPut", args, &reply)
-
+	err := client.Call("ServerClient.ServerPut", args, &reply)
 	if err != nil {
 		fmt.Println(err)
 	} else {
+		// On a successful put, add this transaction into the client's history
 		fmt.Println("Put successful")
+		currTransaction := history{"put", key, value, clientId, serverConn, reply}
+		clientHistory = append(clientHistory, currTransaction)
```

Comment:
Why do we need to store the history in this way? I thought we wanted to store a map of key, clientId versus the logical time of that client for the value of the key.

Reply:
Don't we want to know the clients reads for handling the ERR_DEP in the  ClientGet function? Also we need to know if client has written a value and if read is returning a older value right? So maybe we need the clients reads(gets) and writes(puts).

Correct me if Im wrong though